src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs

```diff
@@ -0,0 +1,246 @@
+            switch (errno)
+            {
+                case 0:
+                    return SocketError.Success;
+                case Interop.libc.HOST_NOT_FOUND:
+                    return SocketError.HostNotFound;
+                case Interop.libc.NO_DATA:
+                    return SocketError.NoData;
+                case Interop.libc.NO_RECOVERY:
+                    return SocketError.NoRecovery;
+                case Interop.libc.TRY_AGAIN:
+                    return SocketError.TryAgain;
+                default:
+                    Debug.Fail("Unexpected errno: " + errno.ToString());
+                    return SocketError.SocketError;
+            }
+        }
+
+        private static SocketError GetSocketErrorForNativeError(int error)
+        {
+            switch (error)
+            {
+                case 0:
+                    return SocketError.Success;
+                case Interop.libc.EAI_AGAIN:
+                    return SocketError.TryAgain;
+                case Interop.libc.EAI_BADFLAGS:
+                    return SocketError.InvalidArgument;
+                case Interop.libc.EAI_FAIL:
+                    return SocketError.NoRecovery;
+                case Interop.libc.EAI_FAMILY:
+                    return SocketError.AddressFamilyNotSupported;
+                case Interop.libc.EAI_NONAME:
+                    return SocketError.HostNotFound;
+                default:
+                    Debug.Fail("Unexpected error: " + error.ToString());
+                    return SocketError.SocketError;
+            }
+        }
+
+        private static unsafe IPHostEntry CreateHostEntry(Interop.libc.hostent* hostent)
+        {
+            string hostName = null;
+            if (hostent->h_name != null)
+            {
+                hostName = Marshal.PtrToStringAnsi((IntPtr)hostent->h_name);
+            }
+
+            int numAddresses;
+            for (numAddresses = 0; hostent->h_addr_list[numAddresses] != null; numAddresses++)
+            {
+            }
+
+            IPAddress[] ipAddresses;
+            if (numAddresses == 0)
+            {
+                ipAddresses = Array.Empty<IPAddress>();
+            }
+            else
+            {
+                ipAddresses = new IPAddress[numAddresses];
+                for (int i = 0; i < numAddresses; i++)
+                {
+                    Debug.Assert(hostent->h_addr_list[i] != null);
+                    ipAddresses[i] = new IPAddress(*(int*)hostent->h_addr_list[i]);
+                }
+            }
+
+            int numAliases;
+            for (numAliases = 0; hostent->h_aliases[numAliases] != null; numAliases++)
+            {
+            }
+
+            string[] aliases;
+            if (numAliases == 0)
+            {
+                aliases = Array.Empty<string>();
+            }
+            else
+            {
+                aliases = new string[numAliases];
+                for (int i = 0; i < numAddresses; i++)
```

Comment:
@pgavlin I have tentatively fixed the code and suggest that we add a UnitTest to avoid such errors. (Tracking this using #2894.)